nirtorch/nir_interpreter.py

```diff
@@ -0,0 +1,285 @@
+        # Recurse into subgraphs
+        if isinstance(node, nir.NIRGraph):
+            owning_module[name] = _construct_module_dict_recursive(node, node_map)
+        else:
+            mapped_module = _map_nir_node_to_torch(node, node_map=node_map)
+            if mapped_module is not None:
+                owning_module[name] = mapped_module
+    return owning_module
+
+
+def _find_input_nodes(
+    name: str,
+    edges: typing.List[typing.Tuple[str, str]],
+    node_outputs: typing.Dict[str, torch.fx.Node],
+):
+    """
+    Looks through the edges and find nodes that are connected to the given node as inputs.
+    If one of the inputs doesn't exist because it hasn't been defined, we return None.
+
+    # TODO: Multiple inputs should be added together
+    """
+    input_node_names = list(filter(lambda t: t[1] == name, edges))
+    output_nodes = tuple(node_outputs.get(a, None) for a, _ in input_node_names)
+    return output_nodes
+
+
+def _is_stateful(module: torch.nn.Module) -> bool:
+    signature = inspect.signature(module.forward)
+    return (
+        "state" in signature.parameters  # Contains a state input argument
+        or isinstance(module, torch.nn.RNNBase)  # RNN modules are stateful
+        or isinstance(
+            module, torch.fx.GraphModule
+        )  # GraphModules created by us are always stateful
+    )
+
+
+def _construct_state_recursive(
+    parent_module: torch.nn.ModuleDict,
+) -> typing.Dict[str, typing.Any]:
+    state = {}
+    for name, module in parent_module.items():
+        if isinstance(module, torch.nn.ModuleDict):
+            state[name] = _construct_state_recursive(module)
+        else:
+            state[name] = None
+    return state
+
+
+def _construct_fx_graph(
+    owning_module: torch.nn.ModuleDict, nir_graph: nir.NIRGraph
+) -> torch.fx.GraphModule:
+    node_outputs = {}
+    recursion_counter = collections.Counter(nir_graph.nodes.keys())
+    torch_graph = torch.fx.Graph(owning_module)
+    # Create a queue of the modules where we can re-insert modules for re-processing
+    # in case of self-reference or if the graph is given out of order
+    module_queue = collections.deque(nir_graph.nodes.items())
+
+    # Create some dummy input for use when processing recursive elements
+    dummy_input = torch_graph.call_function(torch.ones, (1,))
+
+    # Loop through all the nodes in the queue
+    while module_queue:
+        module_name, module = module_queue.popleft()
+        if recursion_counter[module_name] > 3:
+            raise RecursionError(
+                f"Module {module_name} has been traversed multiple times"
+                " which may be a bug in the graph or in the implementation."
+                " Please file an issue at github.com/neuromorphs/nirtorch"
+            )
+
+        if isinstance(module, nir.Input):
+            if len(module.output_type) > 1:
+                raise ValueError("Multiple inputs are currently not supported")
```

Comment:
Would this be better as a NotImplementedError?

Reply:
Yes!